src/main/java/com/faforever/client/fx/contextmenu/ContextMenuBuilder.java

```diff
@@ -0,0 +1,79 @@
+package com.faforever.client.fx.contextmenu;
+
+import javafx.beans.binding.BooleanBinding;
+import javafx.beans.property.BooleanProperty;
+import javafx.scene.control.ContextMenu;
+import javafx.scene.control.SeparatorMenuItem;
+import org.springframework.context.ApplicationContext;
+
+public class ContextMenuBuilder {
+
+  public static Builder newBuilder(ApplicationContext context) {
+    return new Builder(context);
+  }
+
+  public static class Builder {
+
+    private final ApplicationContext applicationContext;
+
+    private final ContextMenu contextMenu = new ContextMenu();
+    private SeparatorMenuItem separator;
+    private BooleanProperty firstItemVisibleProperty;
+    private BooleanBinding totalVisibleBinding;
+
+    private Builder(ApplicationContext applicationContext) {
+      this.applicationContext = applicationContext;
+    }
+
+    public <T, B extends AbstractMenuItem<T>> Builder addItem(Class<B> clazz) {
+      return addItem(clazz, null);
+    }
+
+    public <T, B extends AbstractMenuItem<T>> Builder addItem(Class<B> clazz, T object) {
+      B item = applicationContext.getBean(clazz);
+      item.setObject(object);
+      contextMenu.getItems().add(item);
+      bindVisiblePropertyToSeparator(item.visibleProperty());
+      return this;
+    }
+
+    public <T> Builder addCustomItem(AbstractCustomMenuItemController<T> item) {
+      return addCustomItem(item, null);
+    }
+
+    public <T> Builder addCustomItem(AbstractCustomMenuItemController<T> item, T object) {
+      item.setObject(object);
+      contextMenu.getItems().add(item.getRoot());
+      bindVisiblePropertyToSeparator(item.getRoot().visibleProperty());
+      return this;
+    }
+
+    private void bindVisiblePropertyToSeparator(BooleanProperty visibleProperty) {
+      if (separator != null) {
+        if (firstItemVisibleProperty == null) {
+          separator.visibleProperty().bind(visibleProperty);
+          firstItemVisibleProperty = visibleProperty;
+        } else {
+          if (totalVisibleBinding == null) {
+            totalVisibleBinding = firstItemVisibleProperty.or(visibleProperty);
+          } else {
+            totalVisibleBinding = totalVisibleBinding.or(visibleProperty);
+          }
+          separator.visibleProperty().bind(totalVisibleBinding);
```

Comment:
Why do you need the firstItemVisibleProperty? Couldn't it just be
```suggestion
        if (totalVisibleBinding == null) {
          totalVisibleBinding = Bindings.createBooleanBinding(visibleProperty::get, visibleProperty);
        } else {
          totalVisibleBinding = totalVisibleBinding.or(visibleProperty);
        }
        separator.visibleProperty().unbind();
        separator.visibleProperty().bind(totalVisibleBinding);
```

Reply:
This does not work.

Reply:
```

ContextMenuBuilderTest > testCheckFirstSeparatorVisible() FAILED
    java.lang.ClassCastException: class javafx.beans.property.SimpleBooleanProperty cannot be cast to class javafx.beans.binding.BooleanBinding (javafx.beans.property.SimpleBooleanProperty and javafx.beans.binding.BooleanBinding are in module javafx.base@15.0.1 of loader 'platform')
        at com.faforever.client.fx.contextmenu.ContextMenuBuilder$Builder.bindVisiblePropertyToSeparator(ContextMenuBuilder.java:54)
        at com.faforever.client.fx.contextmenu.ContextMenuBuilder$Builder.addCustomItem(ContextMenuBuilder.java:47)
        at com.faforever.client.fx.contextmenu.ContextMenuBuilder$Builder.addCustomItem(ContextMenuBuilder.java:41)
        at com.faforever.client.fx.contextmenu.ContextMenuBuilderTest.setUp(ContextMenuBuilderTest.java:43)

ContextMenuBuilderTest > testCheckSecondSeparatorVisible() STANDARD_OUT
	


ContextMenuBuilderTest > testCheckSecondSeparatorVisible() FAILED
    java.lang.ClassCastException: class javafx.beans.property.SimpleBooleanProperty cannot be cast to class javafx.beans.binding.BooleanBinding (javafx.beans.property.SimpleBooleanProperty and javafx.beans.binding.BooleanBinding are in module javafx.base@15.0.1 of loader 'platform')
        at com.faforever.client.fx.contextmenu.ContextMenuBuilder$Builder.bindVisiblePropertyToSeparator(ContextMenuBuilder.java:54)
        at com.faforever.client.fx.contextmenu.ContextMenuBuilder$Builder.addCustomItem(ContextMenuBuilder.java:47)
        at com.faforever.client.fx.contextmenu.ContextMenuBuilder$Builder.addCustomItem(ContextMenuBuilder.java:41)
        at com.faforever.client.fx.contextmenu.ContextMenuBuilderTest.setUp(ContextMenuBuilderTest.java:43)
QuantumRenderer: shutdown

Reply:
```suggestion
        if (totalVisibleBinding == null) {
          totalVisibleBinding = Bindings.createBooleanBinding(visibleProperty::get, visibleProperty);
        } else {
          totalVisibleBinding = totalVisibleBinding.or(visibleProperty);
        }
        separator.visibleProperty().unbind();
        separator.visibleProperty().bind(totalVisibleBinding);
```

When I first made the suggestion it had an error with a cast but I just tested this and it does work